## Project Title: Finding Suitable Toronto Neighborhood(s) for Ethnic Food Chain Restaurant

### Upload and Read CSV file containing Postal Codes and Boroughs for Toronto Neighborhoods into a pandas Data Frame df1_Toronto

In [1]:
import pandas as pd

df1_Toronto = pd.read_csv('TorontoZIP.csv')
df1_Toronto.head()

Postal code            Borough                 Neighborhood 
0         M1A       Not assigned                           NaN
1         M2A       Not assigned                           NaN
2         M3A         North York                    Parkwoods 
3         M4A         North York             Victoria Village 
4         M5A   Downtown Toronto   Regent Park / Harbourfront

### Checking for Missing Data in the Data Frame and then Provide a Head of this Missing Data. We will address this missing data a little later where we drop the columns that contain NaN

In [2]:
missing_data = df1_Toronto.isnull()

In [3]:
missing_data.shape

(180, 3)

### First we Rename the Columns of the Above Data Frame ...

In [5]:
df1_Toronto.columns = ['PostalCode','Borough','Neighborhood']
df1_Toronto.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

### Then we strip the column data of any extra whitespaces in front or back ...

In [6]:

df1_Toronto['PostalCode'] = df1_Toronto['PostalCode'].str.strip()
df1_Toronto['Borough'] = df1_Toronto['Borough'].str.strip()
df1_Toronto['Neighborhood'] = df1_Toronto['Neighborhood'].str.strip()

df1_Toronto.dtypes


PostalCode      object
Borough         object
Neighborhood    object
dtype: object

In [7]:
df1_Toronto.head()

PostalCode           Borough                Neighborhood
0        M1A      Not assigned                         NaN
1        M2A      Not assigned                         NaN
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

### In the Borough column we replace NOT ASSIGNED by NaN the PYTHON default for no data ...

In [8]:
import numpy as np
df1_Toronto["Borough"].replace("Not assigned",np.nan,inplace=True)

In [8]:
df1_Toronto.head()

PostalCode           Borough                Neighborhood
0        M1A               NaN                         NaN
1        M2A               NaN                         NaN
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

### Drop rows of Data Frame that have no Boroughs

In [9]:
df1_Toronto.dropna(subset=["Borough"], axis=0, inplace=True)

df1_Toronto.reset_index(drop=True, inplace=True)

In [10]:
df1_Toronto.shape

(103, 3)

### In Neighborhood Column replace forward slash with commas for more desirable format for Neighborhoods

In [11]:
df1_Toronto[["Neighborhood"]] = df1_Toronto[["Neighborhood"]].replace(to_replace=r' / ', value=', ', regex=True)

In [12]:
df1_Toronto.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [13]:
xdim, ydim = df1_Toronto.shape

### Check for any NOT ASSIGNED designations in Neighborhood Column

In [14]:
for i in range(xdim):
    if df1_Toronto.iloc[i,2]=='Not assigned':
        print('There is an Unassigned Neighborhood')

### Upload and Read CSV file containing Postal Codes and Boroughs for Toronto Neighborhoods into a pandas Data Frame df2_Toronto

In [2]:
import pandas as pd
df2_Toronto = pd.read_csv('TorontoGEOSPAT.csv')
df2_Toronto.head()




Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [16]:
df2_Toronto.columns = ['PostalCode','Latitude','Longitude']
df2_Toronto.dtypes

PostalCode     object
Latitude      float64
Longitude     float64
dtype: object

In [17]:
df2_Toronto['PostalCode'] = df2_Toronto['PostalCode'].str.strip()

df2_Toronto.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

### The Latitude and Longitude Data is Ready to Be Merged with the Data Frame of Postal Codes, Boroughs and Neighborhoods created in the prior part.
### First we make a copy of df1_Toronto and call it df3_Toronto so as to keep the former safe for future reference as follows:


In [18]:

df3_Toronto = df1_Toronto.copy()


In [19]:
df3_Toronto.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### Now we use the Pandas DataFrame Merge Function to Merge df3_Toronto with df2_Toronto on the PostalCode Key as below. Then we name the new DataFrame df4_Toronto. Finally we check the shape of df4_Toronto. It should be (103, 5) owing to the new columns of Latitude and Longitude added ...

In [20]:
df4_Toronto = df3_Toronto.merge(df2_Toronto, left_on='PostalCode', right_on='PostalCode')

df4_Toronto.shape

(103, 5)

### Again we make a copy of df4_Toronto and call it dfs_Toronto so we can use it to find the suitable neighborhoods where we should consider opening the new restaurant ...

In [21]:

df4_Toronto.head()

#dfs_Toronto = df4_Toronto.iloc[0:100:10,:]

dfs_Toronto = df4_Toronto.copy()

dfs_Toronto.shape

(103, 5)

### We start with importing and installing the necessary packages

In [22]:
#!pip install geopy
!pip install folium
import pandas as pd

#import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
#from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


### Setting Up FOURSQUARE API Location Data Credentials ...

In [23]:
CLIENT_ID = 'IFDZA4FFK2YR53ZE4AKVGTABRV0HSRQGDMCDWUQGKSF2EQU2' # your Foursquare ID
CLIENT_SECRET = 'MPYD03EF42WT2Z5RR51JCQAELZAD4TPCIX35GVK020OEOXCA' # your Foursquare Secret
VERSION = '20180604' # Foursquare API version
LIMIT = 100
radius = 1000

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: IFDZA4FFK2YR53ZE4AKVGTABRV0HSRQGDMCDWUQGKSF2EQU2
CLIENT_SECRET:MPYD03EF42WT2Z5RR51JCQAELZAD4TPCIX35GVK020OEOXCA


### Establish a Search Query to use in the FOURSQUARE API to look for restaurant venues ... 

In [24]:
search_query = 'Restaurant'
print(search_query + ' .... OK!')


Restaurant .... OK!


### Function getNearybyRestaurants that returns Data Frame with Neighborhood, Latitude, Longitude and Short Name Category for Each Neighborhood ...

In [26]:
def getNearbyRestaurants(names, latitudes, longitudes, radius):
    column_names = ['Neighborhood', 'Latitude', 'Longitude','Category'] 
    TorontoRest = pd.DataFrame(columns = column_names)
    for name, lat, lng in zip(names, latitudes, longitudes):
#        print(name,lat,lng)
#        print(name)           
# create the API request URL
        url='https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(             
        CLIENT_ID,CLIENT_SECRET,lat,lng,VERSION,search_query,radius,LIMIT)
            
        # make the GET request
        results = requests.get(url).json()['response']['venues']
        check = len(results)
#        print('Length of Results = ',check)
#        print(len(results))
#        print(results)
        
        # return only relevant information for each nearby venue
        if (check > 0):
            for v in results:
                check2 = len(v['categories'])
                if (check2 > 0):
#                    print('Length of vcategories = ',len(v['categories']))
#                    print(v['categories'])
                    x = v['categories'][0]['shortName']
                    print('The Short Name is: ',x)
#                x = v['categories'][0]['name']
#                print(v['categories'][0]['shortName'])
                    TorontoRest = TorontoRest.append({'Neighborhood':name,'Latitude':lat,'Longitude':lng,'Category':x},ignore_index=True)
#            venues_list.append([name,lat,lng,v['categories'][0]['shortName']])
#        venues_list.append([(name,lat,lng,v['categories'][0]['shortName']) for v in results])

#    Tor_Rest = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
#    Tor_Rest.columns = ['Neighborhood', 
#                  'Neighborhood Latitude', 
#                  'Neighborhood Longitude', 
#                  'Category']
    
    return(TorontoRest)

### Call The Function Above to Create DataFrame with Neighborhoods, Latitude, Longitude and Category of Restaurant.

### Call the returned data frame Tor_Rest

In [28]:
a = dfs_Toronto['Neighborhood']
b = dfs_Toronto['Latitude']
c = dfs_Toronto['Longitude']

Tor_Rest = getNearbyRestaurants(a,b,c,radius)

The Short Name is:  Caribbean
The Short Name is:  Restaurant
The Short Name is:  Indian
The Short Name is:  American
The Short Name is:  American
The Short Name is:  Ethiopian
The Short Name is:  Breakfast
The Short Name is:  African
The Short Name is:  Asian
The Short Name is:  Food
The Short Name is:  Restaurant
The Short Name is:  Breakfast
The Short Name is:  Diner
The Short Name is:  Restaurant
The Short Name is:  Lounge
The Short Name is:  Restaurant
The Short Name is:  Greek
The Short Name is:  Breakfast
The Short Name is:  Vietnamese
The Short Name is:  Spanish
The Short Name is:  Asian
The Short Name is:  Thai
The Short Name is:  Restaurant
The Short Name is:  Italian
The Short Name is:  Breakfast
The Short Name is:  Sandwiches
The Short Name is:  Restaurant
The Short Name is:  Diner
The Short Name is:  Gastropub
The Short Name is:  Beer Bar
The Short Name is:  Chinese
The Short Name is:  Vietnamese
The Short Name is:  Sushi
The Short Name is:  Breakfast
The Short Name is:  Am

### Finding Shape of Returned Data Frame ...

In [29]:
Tor_Rest.shape


(1365, 4)

## Some Exploratory Data Analytis ...
### 1 . Group Restaurants by Categories to Obtain a Count for Each ...

In [30]:

Tor_Rest.groupby(['Neighborhood']).count()
        

Latitude  Longitude  \
Neighborhood                                                              
Agincourt                                                 20         20   
Alderwood, Long Branch                                     3          3   
Bathurst Manor, Wilson Heights, Downsview North            5          5   
Bedford Park, Lawrence Manor East                          6          6   
Berczy Park                                               50         50   
Birch Cliff, Cliffside West                                3          3   
Brockton, Parkdale Village, Exhibition Place              13         13   
Business reply mail Processing CentrE                      2          2   
CN Tower, King and Spadina, Railway Lands, Harb...         4          4   
Caledonia-Fairbanks                                       10         10   
Canada Post Gateway Processing Centre                      5          5   
Cedarbrae                                                  7          7   
Central Bay Street                                        50         50   
Christie                                                  38         38   
Church and Wellesley                                      50         50   
Clarks Corners, Tam O'Shanter, Sullivan                    8          8   
Cliffside, Cliffcrest, Scarborough Village West            1          1   
Commerce Court, Victoria Hotel                            50         50   
Davisville                                                19         19   
Davisville North                                          17         17   
Del Ray, Mount Dennis, Keelsdale and Silverthorn           5          5   
Don Mills                                                  7          7   
Dorset Park, Wexford Heights, Scarborough Town ...         6          6   
Downsview                                                 15         15   
Dufferin, Dovercourt Village                               9          9   
East Toronto                                              32         32   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...         2          2   
Fairview, Henry Farm, Oriole                               5          5   
First Canadian Place, Underground city                    50         50   
Forest Hill North & West                                   9          9   
Garden District, Ryerson                                  50         50   
Glencairn                                                  6          6   
Golden Mile, Clairlea, Oakridge                            3          3   
Guildwood, Morningside, West Hill                          3          3   
Harbourfront East, Union Station, Toronto Islands         49         49   
High Park, The Junction South                              6          6   
Hillcrest Village                                          3          3   
Humber Summit                                              2          2   
Humewood-Cedarvale                                        10         10   
India Bazaar, The Beaches West                             9          9   
Kennedy Park, Ionview, East Birchmount Park                5          5   
Kensington Market, Chinatown, Grange Park                 49         49   
Kingsview Village, St. Phillips, Martin Grove G...         2          2   
Lawrence Manor, Lawrence Heights                           3          3   
Lawrence Park                                              1          1   
Leaside                                                    2          2   
Little Portugal, Trinity                                  25         25   
Milliken, Agincourt North, Steeles East, L'Amor...        10         10   
Mimico NW, The Queensway West, South of Bloor, ...         2          2   
Moore Park, Summerhill East                               11         11   
New Toronto, Mimico South, Humber Bay Shores               3          3   
North Park, Maple Leaf Park, Upwood Park                   2          2   
No

### 2. Find Unique Restaurant Venue Categories Using df.unique() function from Pandas ...

In [31]:
a = Tor_Rest['Category'].unique()
print("There are %d Unique Venue Categories" % len(a))

There are 101 Unique Venue Categories


### Analyze each Neighborhood for the Restaurant categories via One Hot Encoding. Here we use the pd.get_dummies function to convert categorical data such as restaurant category names to integer numeric data

In [32]:
# One Hot Encoding ...
Tor_Rest_onehot = pd.get_dummies(Tor_Rest[['Category']], prefix="", prefix_sep="")

# Add neighborhood column back to dataframe
Tor_Rest_onehot['Neighborhood'] = Tor_Rest['Neighborhood'] 

# Move neighborhood column to the first column
fixed_columns = [Tor_Rest_onehot.columns[-1]] + list(Tor_Rest_onehot.columns[:-1])
Tor_Rest_onehot = Tor_Rest_onehot[fixed_columns]

Tor_Rest_onehot.shape

Tor_Rest_onehot.head()

Neighborhood  African  American  Argentinian  Asian  Bagels  \
0           Victoria Village        0         0            0      0       0   
1           Victoria Village        0         0            0      0       0   
2           Victoria Village        0         0            0      0       0   
3  Regent Park, Harbourfront        0         1            0      0       0   
4  Regent Park, Harbourfront        0         1            0      0       0   

   Bar  Beer Bar  Bistro  Breakfast  Building  Burgers  Café  Cajun / Creole  \
0    0         0       0          0         0        0     0               0   
1    0         0       0          0         0        0     0               0   
2    0         0       0          0         0        0     0               0   
3    0         0       0          0         0        0     0               0   
4    0         0       0          0         0        0     0               0   

   Cantonese  Caribbean  Chinese  Cocktail  Coworking Space  Deli / Bodega  \
0          0          1        0         0                0              0   
1          0          0        0         0                0              0   
2          0          0        0         0                0              0   
3          0          0        0         0                0              0   
4          0          0        0         0                0              0   

   Department Store  Dim Sum  Diner  Dive Bar  Eastern European  English  \
0                 0        0      0         0                 0        0   
1                 0        0      0         0                 0        0   
2                 0        0      0         0                 0        0   
3                 0        0      0         0                 0        0   
4                 0        0      0         0                 0        0   

   Entertainment  Ethiopian  Event Space  Falafel  Fast Food  Filipino  Food  \
0              0          0            0        0          0         0     0   
1              0          0            0        0          0         0     0   
2              0          0            0        0          0         0     0   
3              0          0            0        0          0         0     0   
4              0          0            0        0          0         0     0   

   French  Furniture / Home  Gastropub  German  Greek  Grocery Store  Gym  \
0       0                 0          0       0      0              0    0   
1       0                 0          0       0      0              0    0   
2       0                 0          0       0      0              0    0   
3       0                 0          0       0      0              0    0   
4       0                 0          0       0      0              0    0   

   Hakka  Halal  Hotel  Hotpot  Hunan  Ice Cream  Indian  Indian Chinese  \
0      0      0      0       0      0          0       0               0   
1      0      0      0       0      0          0       0               0   
2      0      0      0       0      0          0       1               0   
3      0      0      0       0      0          0       0               0   
4      0      0      0       0      0          0       0               0   

   Italian  Japanese  Jazz Club  Juice Bar  Karaoke  Korean  Latin American  \
0        0         0          0          0        0       0               0   
1        0         0          0          0        0       0               0   
2        0         0          0          0        0       0               0   
3        0         0          0          0        0       0               0   
4        0         0          0          0        0       0               0   

   Lebanese  Lounge  Mac & Cheese  Malay  Mediterranean  Mexican  \
0         0       0             0      0              0        0   
1         0       0             0      0              0        0   
2         0       0             0      0              

### Group Rows by Neighborhoods and Take Mean ... Then Reset Index ...

In [34]:
Tor_Rest_grouped = Tor_Rest_onehot.groupby('Neighborhood').mean().reset_index()
Tor_Rest_grouped.head()

Tor_Rest_grouped.shape

#Tor_Rest_grouped.head(5)

(90, 102)

### Function to Sort Restaurant Categories in Descending Order. This is purely illustrative to show the top restaurant venues for each Neighborhood ...

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Get DataFrame with top 20 Restaurant Categories in Each Neighborhood. Again illustrative in nature.

In [36]:
num_top_venues = 20

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
Tor_Rest_venues_sorted = pd.DataFrame(columns=columns)
Tor_Rest_venues_sorted['Neighborhood'] = Tor_Rest_grouped['Neighborhood']

for ind in np.arange(Tor_Rest_grouped.shape[0]):
    Tor_Rest_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Tor_Rest_grouped.iloc[ind, :], num_top_venues)

Tor_Rest_venues_sorted.head()
#neighborhoods_venues_sorted.shape

Neighborhood 1st Most Common Venue  \
0                                        Agincourt               Chinese   
1                           Alderwood, Long Branch                 Pizza   
2  Bathurst Manor, Wilson Heights, Downsview North        Middle Eastern   
3                Bedford Park, Lawrence Manor East            Restaurant   
4                                      Berczy Park            Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Restaurant              Filipino                 Malay   
1                Korean                  Shop                 Wings   
2                 Wings                   Bar            Sandwiches   
3               Italian              Filipino             Breakfast   
4                 Asian             Fast Food                 Diner   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              American                Indian                 Asian   
1                German             Ethiopian           Event Space   
2                 Greek           Event Space               Falafel   
3                 Sushi                German             Ethiopian   
4          New American                Indian              Japanese   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
0           Other Event                Korean             Sandwiches   
1               Falafel             Fast Food               Filipino   
2             Fast Food              Filipino                   Food   
3           Event Space               Falafel              Fast Food   
4            Sandwiches              Wine Bar   Molecular Gastronomy   

  11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
0              Cantonese       Furniture / Home                  Wings   
1                   Food                 French       Furniture / Home   
2                 French       Furniture / Home              Gastropub   
3                   Food                 French       Furniture / Home   
4                   Food          Deli / Bodega                 Lounge   

  14th Most Common Venue 15th Most Common Venue 16th Most Common Venue  \
0                 French                   Food              Fast Food   
1              Gastropub                  Greek                English   
2                 German          Grocery Store          Entertainment   
3              Gastropub          Grocery Store                  Greek   
4                    Pub              Caribbean              Mongolian   

  17th Most Common Venue 18th Most Common Venue 19th Most Common Venue  \
0                Falafel              Gastropub          Grocery Store   
1          Grocery Store                    Gym                  Hakka   
2                    Gym                  Hakka                  Halal   
3                English                    Gym                  Hakka   
4                   Thai               American                Spanish   

  20th Most Common Venue  
0                 German  
1                  Halal  
2                  Hotel  
3                  Halal  
4                  Sushi

### Perform K-Means Clustering on the Data Frame Above. Note this K-Means Machine Learning Algorithm is unsupervised and unlabeled AND it is applied to the Data Frame Toronto_Rest_grouped

In [37]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

Tor_Rest_grouped_clustering = Tor_Rest_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Tor_Rest_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

kmeans.labels_

array([3, 4, 4, 2, 2, 4, 4, 4, 1, 2, 2, 4, 3, 4, 4, 3, 3, 2, 4, 2, 3, 2,
       3, 2, 4, 4, 2, 4, 2, 4, 2, 4, 2, 3, 2, 4, 3, 2, 4, 2, 1, 3, 2, 3,
       1, 0, 2, 3, 4, 4, 4, 4, 2, 4, 2, 2, 4, 3, 3, 2, 2, 4, 2, 4, 2, 4,
       4, 2, 2, 4, 3, 2, 2, 4, 4, 4, 4, 0, 3, 2, 3, 2, 2, 4, 4, 4, 4, 2,
       4, 0])

### So now we have clustered our neighborhoods based on the restaurant venues around them. Next we will use Folium to create a cluster map. To this end we need some more analysis as below.

In [38]:
# add clustering labels
Tor_Rest_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Tor_Rest_merged = dfs_Toronto.copy()

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Tor_Rest_merged = Tor_Rest_merged.join(Tor_Rest_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Tor_Rest_merged.head() 



PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             NaN                   NaN   
1  43.725882 -79.315572             2.0                Indian   
2  43.654260 -79.360636             2.0            Restaurant   
3  43.718518 -79.464763             3.0            Vietnamese   
4  43.662301 -79.389494             3.0               Chinese   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   NaN                   NaN                   NaN   
1            Restaurant             Caribbean                 Wings   
2             Breakfast                 Diner              American   
3               Chinese                 Sushi                 Wings   
4                 Asian                Indian            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   NaN                   NaN                   NaN   
1                German           Event Space               Falafel   
2                 Asian               Spanish                Lounge   
3                German             Ethiopian           Event Space   
4              American               Italian                  Thai   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
0                   NaN                   NaN                    NaN   
1             Fast Food              Filipino                   Food   
2                 Greek             Gastropub                   Food   
3               Falafel             Fast Food               Filipino   
4              Szechuan          New American              Caribbean   

  11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
0                    NaN                    NaN                    NaN   
1                 French       Furniture / Home              Gastropub   
2              Ethiopian             Sandwiches                African   
3                   Food                 French       Furniture / Home   
4        Modern European              Gastropub               Cocktail   

  14th Most Common Venue 15th Most Common Venue 16th Most Common Venue  \
0                    NaN                    NaN                    NaN   
1                  Greek          Entertainment          Grocery Store   
2                Italian               Beer Bar                   Thai   
3              Gastropub          Grocery Store                  Greek   
4                 Korean                 Bistro              Breakfast   

  17th Most Common Venue 18th Most Common Venue 19th Most Common Venue  \
0                    NaN                    NaN                    NaN   
1                    Gym                  Hakka                  Halal   
2             Vietnamese                  Hunan            Argentinian   
3                English                    Gym                  Hakka   
4                Dim Sum               Japanese       Theme Restaurant   

  20th Most Common Venue  
0                    NaN  
1                  Hotel  
2                 Indian  
3                  Halal  
4            Event Space

### K-Means Could Not Include All Neighborhoods into Clusters and hence some of the output was a NaN. Therefore we drop those rows and convert the Cluster Labels Column to Integer so that it can be mapped using Folium in the next Cell

In [39]:

Tor_Rest_merged.dropna(subset=["Cluster Labels"], axis=0, inplace=True)
Tor_Rest_merged['Cluster Labels'] = Tor_Rest_merged['Cluster Labels'].astype('int')
Tor_Rest_merged.head()

PostalCode           Borough                                 Neighborhood  \
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
7        M3B        North York                                    Don Mills   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
1  43.725882 -79.315572               2                Indian   
2  43.654260 -79.360636               2            Restaurant   
3  43.718518 -79.464763               3            Vietnamese   
4  43.662301 -79.389494               3               Chinese   
7  43.745906 -79.352188               2               Italian   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1            Restaurant             Caribbean                 Wings   
2             Breakfast                 Diner              American   
3               Chinese                 Sushi                 Wings   
4                 Asian                Indian            Restaurant   
7              Vineyard              American                Indian   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1                German           Event Space               Falafel   
2                 Asian               Spanish                Lounge   
3                German             Ethiopian           Event Space   
4              American               Italian                  Thai   
7                 Diner            Restaurant                   Gym   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
1             Fast Food              Filipino                   Food   
2                 Greek             Gastropub                   Food   
3               Falafel             Fast Food               Filipino   
4              Szechuan          New American              Caribbean   
7             Gastropub           Event Space                Falafel   

  11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
1                 French       Furniture / Home              Gastropub   
2              Ethiopian             Sandwiches                African   
3                   Food                 French       Furniture / Home   
4        Modern European              Gastropub               Cocktail   
7              Fast Food               Filipino                   Food   

  14th Most Common Venue 15th Most Common Venue 16th Most Common Venue  \
1                  Greek          Entertainment          Grocery Store   
2                Italian               Beer Bar                   Thai   
3              Gastropub          Grocery Store                  Greek   
4                 Korean                 Bistro              Breakfast   
7                 French       Furniture / Home          Grocery Store   

  17th Most Common Venue 18th Most Common Venue 19th Most Common Venue  \
1                    Gym                  Hakka                  Halal   
2             Vietnamese                  Hunan            Argentinian   
3                English                    Gym                  Hakka   
4                Dim Sum               Japanese       Theme Restaurant   
7                 German                  Greek          Entertainment   

  20th Most Common Venue  
1                  Hotel  
2                 Indian  
3                  Halal  
4            Event Space  
7                  Hakka

### Next we use Folium to create a map. The map is saved to be accesed.

In [41]:
# create map

latitude = 43.6487
longitude = -79.38544

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Tor_Rest_merged['Latitude'], Tor_Rest_merged['Longitude'], Tor_Rest_merged['Neighborhood'], Tor_Rest_merged['Cluster Labels']):    
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters.save('F35.html')

### We now examine each cluster to determine the discriminating venue categories. That way we can try to understand our clusters using the categories therein

### CLUSTER - 1

In [49]:
#Tor_Rest_merged.loc[Tor_Rest_merged['Cluster Labels'] == 0, Tor_Rest_merged.columns[[1] + list(range(5, Tor_Rest_merged.shape[1]))]]

Cluster1 = Tor_Rest_merged.loc[Tor_Rest_merged['Cluster Labels'] == 0,:]

Cluster1

PostalCode     Borough                                   Neighborhood  \
23        M4G   East York                                        Leaside   
66        M2P  North York                                York Mills West   
98        M8X   Etobicoke  The Kingsway, Montgomery Road, Old Mill North   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
23  43.709060 -79.363452               0                Indian   
66  43.752758 -79.400049               0                 Diner   
98  43.653654 -79.506944               0              American   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
23                 Sushi                 Wings                 Greek   
66                 Sushi                 Wings                 Greek   
98                 Sushi                 Wings         Grocery Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23             Ethiopian           Event Space               Falafel   
66           Event Space               Falafel             Fast Food   
98           Event Space               Falafel             Fast Food   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
23             Fast Food              Filipino                   Food   
66              Filipino                  Food                 French   
98              Filipino                  Food                 French   

   11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
23                 French       Furniture / Home              Gastropub   
66       Furniture / Home              Gastropub                 German   
98       Furniture / Home              Gastropub                 German   

   14th Most Common Venue 15th Most Common Venue 16th Most Common Venue  \
23                 German          Grocery Store                English   
66          Grocery Store          Entertainment                    Gym   
98                  Greek                    Gym          Entertainment   

   17th Most Common Venue 18th Most Common Venue 19th Most Common Venue  \
23                    Gym                  Hakka                  Halal   
66                  Hakka                  Halal                  Hotel   
98                  Hakka                  Halal                  Hotel   

   20th Most Common Venue  
23                  Hotel  
66                 Hotpot  
98                 Hotpot

### CLUSTER - 2

In [50]:
##Tor_Rest_merged.loc[Tor_Rest_merged['Cluster Labels'] == 1, Tor_Rest_merged.columns[[1] + list(range(5, Tor_Rest_merged.shape[1]))]]

Cluster2 = Tor_Rest_merged.loc[Tor_Rest_merged['Cluster Labels'] == 1,:]

Cluster2

PostalCode           Borough  \
38        M1K       Scarborough   
61        M4N   Central Toronto   
87        M5V  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
38        Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029   
61                                      Lawrence Park  43.728020 -79.388790   
87  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
38               1                 Asian                   Bar   
61               1                   Bar                 Wings   
87               1      Eastern European                 Asian   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
38            Restaurant               Chinese                 Wings   
61         Grocery Store           Event Space               Falafel   
87                   Bar               Chinese                 Wings   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
38                 Greek               Falafel             Fast Food   
61             Fast Food              Filipino                  Food   
87                 Greek               Falafel             Fast Food   

   9th Most Common Venue 10th Most Common Venue 11th Most Common Venue  \
38              Filipino                   Food                 French   
61                French       Furniture / Home              Gastropub   
87              Filipino                   Food                 French   

   12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
38       Furniture / Home              Gastropub                 German   
61                 German                  Greek                    Gym   
87       Furniture / Home              Gastropub                 German   

   15th Most Common Venue 16th Most Common Venue 17th Most Common Venue  \
38                    Gym          Grocery Store              Ethiopian   
61          Entertainment                  Hakka                  Halal   
87                    Gym          Grocery Store              Ethiopian   

   18th Most Common Venue 19th Most Common Venue 20th Most Common Venue  
38                  Hakka                  Halal                  Hotel  
61                  Hotel                 Hotpot                  Hunan  
87                  Hakka                  Halal                  Hotel

### CLUSTER - 3

In [51]:
#Tor_Rest_merged.loc[Tor_Rest_merged['Cluster Labels'] == 2, Tor_Rest_merged.columns[[1] + list(range(5, Tor_Rest_merged.shape[1]))]]

Cluster3 = Tor_Rest_merged.loc[Tor_Rest_merged['Cluster Labels'] == 2,:]

Cluster3



PostalCode           Borough  \
1          M4A        North York   
2          M5A  Downtown Toronto   
7          M3B        North York   
9          M5B  Downtown Toronto   
13         M3C        North York   
15         M5C  Downtown Toronto   
17         M9C         Etobicoke   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
30         M5H  Downtown Toronto   
34         M3J        North York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
40         M3K        North York   
42         M5K  Downtown Toronto   
44         M1L       Scarborough   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
50         M9L        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
60         M3N        North York   
62         M5N   Central Toronto   
67         M4P   Central Toronto   
70         M9P         Etobicoke   
73         M4R   Central Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
81         M6S      West Toronto   
89         M9V         Etobicoke   
92         M5W  Downtown Toronto   
97         M5X  Downtown Toronto   
101        M8Y         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
1                                     Victoria Village  43.725882 -79.315572   
2                            Regent Park, Harbourfront  43.654260 -79.360636   
7                                            Don Mills  43.745906 -79.352188   
9                             Garden District, Ryerson  43.657162 -79.378937   
13                                           Don Mills  43.725900 -79.340923   
15                                      St. James Town  43.651494 -79.375418   
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201   
20                                         Berczy Park  43.644771 -79.373306   
21                                 Caledonia-Fairbanks  43.689026 -79.453512   
22                                              Woburn  43.770992 -79.216917   
30                            Richmond, Adelaide, King  43.650571 -79.384568   
34                     Northwood Park, York University  43.767980 -79.487262   
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
37                            Little Portugal, Trinity  43.647927 -79.419750   
40                                           Downsview  43.737473 -79.464763   
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
44                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577   
46                                           Downsview  43.739015 -79.506944   
47                      India Bazaar, The Beaches West  43.668999 -79.315572   
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817   
50                                       Humber Summit  43.756303 -79.565963   
53                                           Downsview  43.728496 -79.495697   
54                                     Studio District  43.659526 -79.340923   
55                   Bedford Park, Lawrence Manor East  43.733283 -79.419750   
60                                           Downsview  43.761631 -79.520999   
62                                            Roselawn  43.711695 -79.416936   
67                                    Davisville North  43.712751 -79.390197   
70                                           Westmount  43.696319 -79.532242   
73                                  North Toronto West  43.715383 -79.405678   
76               Canada Post Gateway Processing Centre  43.636966 -79.615819   
77   Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724   
81                                  Runnymede, Swansea  43.651571 -79.484450   
89   South Steeles, Silverstone

### CLUSTER - 4

In [52]:
#Tor_Rest_merged.loc[Tor_Rest_merged['Cluster Labels'] == 3, Tor_Rest_merged.columns[[1] + list(range(5, Tor_Rest_merged.shape[1]))]]

Cluster4 = Tor_Rest_merged.loc[Tor_Rest_merged['Cluster Labels'] == 3,:]

Cluster4



PostalCode           Borough  \
3         M6A        North York   
4         M7A  Downtown Toronto   
8         M4B         East York   
18        M1E       Scarborough   
24        M5G  Downtown Toronto   
27        M2H        North York   
29        M4H         East York   
51        M1M       Scarborough   
56        M6M              York   
65        M1P       Scarborough   
78        M1S       Scarborough   
80        M5S  Downtown Toronto   
82        M1T       Scarborough   
84        M5T  Downtown Toronto   
85        M1V       Scarborough   
90        M1W       Scarborough   

                                         Neighborhood   Latitude  Longitude  \
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937   
18                  Guildwood, Morningside, West Hill  43.763573 -79.188711   
24                                 Central Bay Street  43.657952 -79.387383   
27                                  Hillcrest Village  43.803762 -79.363452   
29                                   Thorncliffe Park  43.705369 -79.349372   
51    Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476   
56   Del Ray, Mount Dennis, Keelsdale and Silverthorn  43.691116 -79.476013   
65  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304   
78                                          Agincourt  43.794200 -79.262029   
80                     University of Toronto, Harbord  43.662696 -79.400049   
82            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302   
84          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
85  Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577   
90                      Steeles West, L'Amoreaux West  43.799525 -79.318389   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
3                3            Vietnamese               Chinese   
4                3               Chinese                 Asian   
8                3              American                Indian   
18               3               Chinese             Fast Food   
24               3               Chinese            Restaurant   
27               3              Japanese            Sandwiches   
29               3               Chinese               African   
51               3               Chinese                 Wings   
56               3               Chinese              American   
65               3               Chinese              American   
78               3               Chinese            Restaurant   
80               3               Chinese                Korean   
82               3               Chinese                Korean   
84               3               Chinese            Vietnamese   
85               3               Chinese    Vegetarian / Vegan   
90               3               Chinese            Vietnamese   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
3                  Sushi                 Wings                German   
4                 Indian            Restaurant              American   
8                Burgers               Chinese                 Wings   
18                 Wings                 Greek           Event Space   
24               Dim Sum                Indian                Korean   
27               Chinese         Entertainment           Event Space   
29                Indian               Turkish             Caribbean   
51                 Greek           Event Space               Falafel   
56            Portuguese                 Sushi                 Wings   
65                Indian              Filipino        Middle Eastern   
78              Filipino                 Malay              American   
80                 Asian            Restaurant            Vietnamese   
82  

### CLUSTER - 5

In [55]:
#Tor_Rest_merged.loc[Tor_Rest_merged['Cluster Labels'] == 4, Tor_Rest_merged.columns[[1] + list(range(5, Tor_Rest_merged.shape[1]))]]
Cluster5 = Tor_Rest_merged.loc[Tor_Rest_merged['Cluster Labels'] == 4,:]

Cluster5


PostalCode           Borough  \
10         M6B        North York   
12         M1C       Scarborough   
14         M4C         East York   
16         M6C              York   
19         M4E      East Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
28         M3H        North York   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
35         M4J         East York   
41         M4K      East Toronto   
43         M6K      West Toronto   
49         M6L        North York   
52         M2M        North York   
58         M1N       Scarborough   
59         M2N        North York   
63         M6N              York   
64         M9N              York   
68         M5P   Central Toronto   
69         M6P      West Toronto   
71         M1R       Scarborough   
72         M2R        North York   
74         M5R   Central Toronto   
75         M6R      West Toronto   
79         M4S   Central Toronto   
83         M4T   Central Toronto   
86         M4V   Central Toronto   
88         M8V         Etobicoke   
91         M4W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
96         M4X  Downtown Toronto   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
10                                           Glencairn  43.709577 -79.445073   
12              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
14                                    Woodbine Heights  43.695344 -79.318389   
16                                  Humewood-Cedarvale  43.693781 -79.428191   
19                                         The Beaches  43.676357 -79.293031   
25                                            Christie  43.669542 -79.422564   
26                                           Cedarbrae  43.773136 -79.239476   
28     Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259   
31                        Dufferin, Dovercourt Village  43.669005 -79.442259   
32                                 Scarborough Village  43.744734 -79.239476   
33                        Fairview, Henry Farm, Oriole  43.778517 -79.346556   
35                                        East Toronto  43.685347 -79.338106   
41                        The Danforth West, Riverdale  43.679557 -79.352188   
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
49            North Park, Maple Leaf Park, Upwood Park  43.713756 -79.490074   
52                             Willowdale, Newtonbrook  43.789053 -79.408493   
58                         Birch Cliff, Cliffside West  43.692657 -79.264848   
59                                          Willowdale  43.770120 -79.408493   
63                       Runnymede, The Junction North  43.673185 -79.487262   
64                                              Weston  43.706876 -79.518188   
68                            Forest Hill North & West  43.696948 -79.411307   
69                       High Park, The Junction South  43.661608 -79.464763   
71                                   Wexford, Maryvale  43.750072 -79.295849   
72                                          Willowdale  43.782736 -79.442259   
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678   
75                              Parkdale, Roncesvalles  43.648960 -79.456325   
79                                          Davisville  43.704324 -79.388790   
83                         Moore Park, Summerhill East  43.689574 -79.383160   
86   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
88        New Toronto, Mimico South, Humber Bay Shores  43.605647 -79.501321   
91                                            Rosedale  43.679563 -79.377529   
93                              Alderwood, Long Branch  43.602414 -79.54348